docs/sources/mimir/references/architecture/mimir-query-engine.md

```diff
@@ -0,0 +1,101 @@
+engine for queries that use unsupported features.
+
+## How to enable MQE
+
+MQE is experimental and disabled by default. It can be enabled by setting the
+`-querier.query-engine=mimir` CLI flag on queriers, or setting the equivalent YAML
+configuration file option.
+
+## Fallback to Prometheus' engine
+
+By default, MQE will fall back to Prometheus' engine for any queries that use unsupported
+features.
+
+This behaviour can be disabled by setting the `-querier.enable-query-engine-fallback=false`
+CLI flag on queriers, or setting the equivalent YAML configuration file option. If fallback
+is disabled and MQE receives a query it does not support, then the query will fail.
+
+You can force a query supported by MQE to use Prometheus' engine by adding the
+`X-Mimir-Force-Prometheus-Engine: true` HTTP header to the query request. This header only
+has an effect if fallback is enabled.
+
+## Query memory consumption limit
+
+MQE supports enforcing a per-query memory consumption limit. This allows you to ensure that
+a single memory-hungry query cannot monopolise a large proportion of available memory in a
+querier, or cause it to exhaust all available memory and crash.
+
+While evaluating a query, MQE estimates the memory consumed by the query, such as memory used
+for the final result and any intermediate calculations, and aborts the query with an
+[`err-mimir-max-estimated-memory-consumption-per-query`]({{< relref "../../manage/mimir-runbooks#err-mimir-max-estimated-memory-consumption-per-query" >}})
+error if the estimate exceeds the configured limit.
+
+The estimate is based on the memory consumed by samples currently held in memory for query
+evaluation. This includes both raw samples decoded from chunks and samples held in memory as
+intermediate results of calculations, or the final result. It also includes some other large
+sources of memory consumption for intermediate results.
+
+The estimate is limited, for example:
+
+- it does not consider the memory consumed by series labels
+- it does not consider the memory consumed by chunks that are currently in memory (although
+  the max chunks and max chunks bytes limits continue to be enforced)
+- it makes an assumption about the memory consumed by each native histogram, rather than
+  accurately calculating the memory consumed by each histogram
+
+By default, no limit is enforced. The default limit for all tenants can be configured by
+setting the `-querier.max-estimated-memory-consumption-per-query` CLI flag, or setting the
+equivalent YAML configuration file option. The default limit can be overridden on a per-tenant
+basis by setting `max_estimated_memory_consumption_per_query` for that tenant. Setting the
+limit to 0 disables it.
+
+The limit is not enforced for queries that run through Prometheus' engine, and setting the limit
+has no impact if MQE is disabled or the query falls back to Prometheus' engine.
+
+## Known differences compared to Prometheus' engine
+
```

Comment:
We should add some introductory text here, such as 

`The following are known differences between MQE and Prometheus' engine:`